server/auth.go

```diff
@@ -8,30 +8,126 @@ import (
+		ctx := context.Background()
+		var user *api.Principal
+
+		authProvider := api.PrincipalAuthProvider(c.Param("auth_provider"))
+		switch authProvider {
+		case api.PrincipalAuthProviderBytebase:
+			{
+				login := &api.Login{}
+				if err := jsonapi.UnmarshalPayload(c.Request().Body, login); err != nil {
+					return echo.NewHTTPError(http.StatusBadRequest, "Malformatted login request").SetInternal(err)
+				}
+
+				principalFind := &api.PrincipalFind{
+					Email: &login.Email,
+				}
+				var err error
+				user, err = s.PrincipalService.FindPrincipal(ctx, principalFind)
+				if err != nil {
+					return echo.NewHTTPError(http.StatusInternalServerError, "Failed to authenticate user").SetInternal(err)
+				}
+				if user == nil {
+					return echo.NewHTTPError(http.StatusUnauthorized, fmt.Sprintf("User not found: %s", login.Email))
+				}
+				// if user had already use third party method logged in, we will forbid she login via Bytebase
+				if user.AuthProvider != api.PrincipalAuthProviderBytebase {
+					return echo.NewHTTPError(http.StatusUnauthorized, "User has bond to Gitlab")
+				}
+				// Compare the stored hashed password, with the hashed version of the password that was received.
+				if err := bcrypt.CompareHashAndPassword([]byte(user.PasswordHash), []byte(login.Password)); err != nil {
+					// If the two passwords don't match, return a 401 status.
+					return echo.NewHTTPError(http.StatusUnauthorized, "Incorrect password").SetInternal(err)
+				}
+			}
+		case api.PrincipalAuthProviderGitlabSelfHost:
+			{
+				gitlabLogin := &api.GitlabLogin{}
+				if err := jsonapi.UnmarshalPayload(c.Request().Body, gitlabLogin); err != nil {
+					return echo.NewHTTPError(http.StatusBadRequest, "Malformatted gitlab login request").SetInternal(err)
+				}
+				GitlabUserInfo, err := vcsPlugin.Get("GITLAB_SELF_HOST", vcsPlugin.ProviderConfig{Logger: s.l}).TryLogin(ctx,
+					common.OauthContext{
+						ClientID:     gitlabLogin.ApplicationID,
+						ClientSecret: gitlabLogin.Secret,
+						AccessToken:  gitlabLogin.AccessToken,
+						RefreshToken: "",
+						Refresher:    nil,
+					},
+					gitlabLogin.InstanceURL,
+				)
+				if err != nil {
+					return echo.NewHTTPError(http.StatusInternalServerError, "Fail to fetch user info from gitlab").SetInternal(err)
+				}
+
+				// we only allow active user to login via gitlab
+				if GitlabUserInfo.State != vcsPlugin.UserStateActive {
+					return echo.NewHTTPError(http.StatusInternalServerError, "Fail to login via Gitlab, user is Archived")
+				}
+
+				principalFind := &api.PrincipalFind{
+					Email: &GitlabUserInfo.Email,
+				}
+				user, err = s.PrincipalService.FindPrincipal(ctx, principalFind)
+				if err != nil {
+					return echo.NewHTTPError(http.StatusInternalServerError, "Failed to authenticate user").SetInternal(err)
+				}
+
+				// create a new user if not exist
+				if user == nil {
+					// if user login via gitlab at the first time, we will generate a random password
```

Comment:
if user login via gitlab at the first time, we will generate a random password. The random password is supposed to be not guessable. But even one can guess the password, she won't be allowed to login. If user signs up via 3rd party login like GitLab, we will disallow bytebase builtin password/email login unless user manually switch the login option later.